shapely/strtree.py

```diff
@@ -213,24 +262,71 @@ def nearest(self, geom):
         >>> tree = STRtree ([Point(0, 0), Point(0, 0)])
         >>> tree.nearest(Point(0, 0)).wkt
         'POINT (0 0)'
+
         """
-        if self._n_geoms == 0:
+        if self._tree is None or not self._rev:
             return None
 
         envelope = geom.envelope
 
         def callback(item1, item2, distance, userdata):
             try:
-                geom1 = ctypes.cast(item1, ctypes.py_object).value
+                idx = ctypes.cast(item1, ctypes.py_object).value
                 geom2 = ctypes.cast(item2, ctypes.py_object).value
                 dist = ctypes.cast(distance, ctypes.POINTER(ctypes.c_double))
-                lgeos.GEOSDistance(geom1._geom, geom2._geom, dist)
+                lgeos.GEOSDistance(self._rev[idx]._geom, geom2._geom, dist)
                 return 1
             except Exception:
+                log.exception("Caught exception")
                 return 0
 
-        item = lgeos.GEOSSTRtree_nearest_generic(self._tree_handle, ctypes.py_object(geom), envelope._geom, \
-            lgeos.GEOSDistanceCallback(callback), None)
+        item = lgeos.GEOSSTRtree_nearest_generic(
+            self._tree,
+            ctypes.py_object(geom),
+            envelope._geom,
+            lgeos.GEOSDistanceCallback(callback),
+            None,
+        )
         result = ctypes.cast(item, ctypes.py_object).value
-
         return result
+
+    def nearest_geom(self, geom):
```

Comment:
If we were to use type hints, this would be something like

`def nearest_geom(self, geom: Geometry) -> Geometry:`

I don't know how to write the type hints for 

`def nearest(self, geom: Geometry, return_geometry: bool = True) -> "Geometry if return_geometry else int":`

I'm not sure if Python supports conditional typing at all right now.

Reply:
Python supports this with the `@overload` decorator (if we would want it): https://docs.python.org/3/library/typing.html#typing.overload